File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
     priority_queue<pair<int,ListNode*>,vector<pair<int,ListNode*>>,greater<pair<int,ListNode*>>> pq;
     ListNode* temp = new ListNode();
     ListNode* ans = temp;
     for( ListNode* t : lists)
     {
        if(t!=NULL) pq.push(make_pair(t->val,t));
     }
     while(!pq.empty())
     {
        ListNode* tt = pq.top().second;
        pq.pop();
        temp->next = tt;
        temp = temp->next;
        if(tt->next!=NULL) pq.push(make_pair(tt->next->val,tt->next));

     }
      return ans->next;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b)->{return a.val-b.val;});
        for(ListNode temp : lists)
        {
            if(temp!=null) pq.add(temp);
        }
        ListNode dummy = new ListNode(0);
        ListNode ans = dummy;
        while(!pq.isEmpty())
        {
             ListNode t = pq.poll();
             ans.next = t;
             ans = ans.next;
             if(t.next!=null) pq.add(t.next);
        }
        return dummy.next;
        
    }
}